trac-34747 To Master : Add Filter For Customizer Description

kienstra opened the pull request:
This pull request for [trac-34747](https://core.trac.wordpress.org/ticket/34747) adds a filter for the Customizer description. This appears at the top of the Customizer's primary panel, after the title.

If there's no description, it doesn't echo the `?` button, which displays it.

Please don't merge this PR, as the "master" branch needs to be in sync with the remote [WordPress repo](https://core.trac.wordpress.org/browser/trunk).

kienstra commented:
Sorry, I'm resolving the merge conflicts now.

westonruter commented:
I forgive you. :smile:

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -184,6 +184,33 @@
+	 * @access protected
+	 * @var string
+	 */
+	protected $filtered_root_panel_title;
```
@kienstra I think this should instead be `$root_panel_title`, and its initial value can be set at the top in `wp-admin/customize.php` to be `get_bloginfo( 'name' )`.

westonruter replied:
On second thought, how about this…

Let's have `$root_panel_title` remain `null` and not get set explicitly…

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -184,6 +184,33 @@
+	 * @access protected
+	 * @var string
+	 */
+	protected $original_customizer_title;
```
I think this can be removed, just letting `$root_panel_title` be `null` initially, and defined explicitly in the top of `wp-admin/customize.php` as is done for setting the `$preview_url`, `$return_url`, and `$autofocus`.

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -184,6 +184,33 @@
+	 * @access protected
+	 * @var string
+	 */
+	protected $filtered_root_panel_description;
```
Per above, let's have this just be `$root_panel_description`

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -1601,6 +1628,93 @@ public function get_return_url() {
+	 * @since 4.6
+	 * @access public
+	 */
+	public function set_filtered_root_panel_title() {
```
Rename this to `set_root_panel_title()`

westonruter replied:
Also, it should accept a `$root_panel_title` as its argument.

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -1601,6 +1628,93 @@ public function get_return_url() {
+	 * @access public
+	 */
+	public function set_filtered_root_panel_title() {
+		$this->original_customizer_title = get_bloginfo( 'name' );
```
This can be removed.

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -1601,6 +1628,93 @@ public function get_return_url() {
+		 *
+		 * @param string $customizer_title Appears at top of root Customizer panel.
+		 */
+		$this->filtered_root_panel_title = apply_filters( 'customize_root_panel_title', $this->original_customizer_title );
```
Let's filter in the getter, not in the setter. So this method can be merely:

```php
$this->root_panel_title = $root_panel_title;
```

Where `$root_panel_title` is passed in as a method argument.

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -1601,6 +1628,93 @@ public function get_return_url() {
+	 * @since 4.6
+	 * @access public
+	 */
+	public function get_filtered_root_panel_title() {
```
Rename to `get_root_panel_title()`.

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -1601,6 +1628,93 @@ public function get_return_url() {
+	public function get_filtered_root_panel_title() {
+		if ( isset( $this->filtered_root_panel_title ) ) {
+			return $this->filtered_root_panel_title;
+		}
```
Replace this conditional with (along with the necessary phpdoc):

```php
return apply_filters( 'customize_root_panel_title', $this->root_panel_title );
```

westonruter replied:
Actually, how about this instead:

```php
return is_null( $this->root_panel_title ) ? get_bloginfo( 'name' ) : $this->root_panel_title;
```

So then I think we can actually eliminate the `customize_root_panel_title` filter entirely. Plugins can just call the method if they want to override the title instead of using a filter.

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -1601,6 +1628,93 @@ public function get_return_url() {
+	 *
+	 * @return bool
+	 */
+	public function is_root_panel_title_filtered() {
```
I see now why you have `original_customizer_title` so that you can pass back whether the title was filtered, to short-circuit the `blogname` updates also updating the root panel title.

Hummmm…

westonruter replied:
Yeah, per my other comments, I think we can use `isset( $this->root_panel_title )` to determine if a custom title has been supplied. If not, the getter will return the `get_bloginfo( 'name' )`

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -1727,6 +1845,9 @@ public function customize_pane_settings() {
 			'previewableDevices' => $this->get_previewable_devices(),
 			'selectiveRefreshEnabled' => isset( $this->selective_refresh ),
+			'filteredRootPanelTitle' => $this->get_filtered_root_panel_title(),
+			'isRootPanelTitleFiltered' => $this->is_root_panel_title_filtered(),
```
How about replacing this with:

```
`isRootPanelTitleOverridden` => ! is_null( $this->root_panel_title ),
```

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -1727,6 +1845,9 @@ public function customize_pane_settings() {
 			'documentTitleTmpl' => $this->get_document_title_template(),
 			'previewableDevices' => $this->get_previewable_devices(),
 			'selectiveRefreshEnabled' => isset( $this->selective_refresh ),
+			'filteredRootPanelTitle' => $this->get_filtered_root_panel_title(),
```
We could leave `$this->root_panel_title` as `null`, but the `get_root_panel_title()` method can check if `$this->root_panel_title` is `null`, and if so return `get_bloginfo( 'name' )`.

kienstra commented:
Hi @westonruter,
Thanks a lot for reviewing this PR. The latest commits apply your suggestions.

Though you didn't suggest changing `set_panel_description`, I [removed the filter](https://github.com/xwp/wordpress-develop/commit/5872ca49da2a4288ebf115cc0a39b8136138b20d#diff-4fb8a477f559bdfad2c1e6db6d1c8b06L1653) in this. This can be changed by passing a new description as an argument to [$wp_customize->set_panel_description](https://github.com/xwp/wordpress-develop/pull/143/files#diff-4fb8a477f559bdfad2c1e6db6d1c8b06R1655).

This is like you suggested for the [set_panel_title](https://github.com/xwp/wordpress-develop/commit/ded3b2c0a4b4e39466d38992b909ca01a459f400) method.

westonruter commented:
@kienstra I was hoping you would infer that. Good on ya!

kienstra commented:
Thanks, @westonruter :smiley:

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -1601,6 +1619,57 @@ public function get_return_url() {
+	 * Appears above the description.
+	 *
+	 * @since 4.6
+	 * @access public
```
Trivial thing: needs a `@return`.

westonruter commented on src/wp-includes/class-wp-customize-manager.php:
```diff
@@ -1601,6 +1619,57 @@ public function get_return_url() {
+	 * Appears after "You are customizing" and the Customizer title.
+	 *
+	 * @since 4.6
+	 * @access public
```
Trivial thing: needs a `@return`.

westonruter commented on src/wp-admin/customize.php:
```diff
@@ -136,9 +136,9 @@
-				?></div>
+				<div class="customize-panel-description">
+					<?php // Description produced with JavaScript ?>
+				</div>
```
Curious as to the reason for using JS to populate this. Since it won't be later updated, should it just `echo $wp_customize->get_root_panel_description()`?

westonruter replied:
Likewise, on line 137 above, this should be changed as:

```diff
- echo sprintf( __( 'You are customizing %s' ), '<strong class="panel-title site-title">' . get_bloginfo( 'name' ) . '</strong>' );
+ echo sprintf( __( 'You are customizing %s' ), '<strong class="panel-title site-title">' . $wp_customize->get_root_panel_title() . '</strong>' ); 
```

kienstra commented:
Hi @westonruter,
Thanks for reviewing these commits. As you suggested, I moved the output of the description to [customize.php](https://github.com/xwp/wordpress-develop/commit/d7c730170770bcea99221d5679fbd12015569214#diff-a325cc1f63ecc04fdb078079c6ec585bR145). And I [removed](https://github.com/xwp/wordpress-develop/commit/d7c730170770bcea99221d5679fbd12015569214#diff-58dfe22165ffd9650354f7bdd51002ecL3253) the JS that added this. 

Likewise, I moved the logic that hides the `.customize-help-toggle` button from [customize-controls.js](https://github.com/xwp/wordpress-develop/commit/d7c730170770bcea99221d5679fbd12015569214#diff-58dfe22165ffd9650354f7bdd51002ecL3257) to [customize.php](https://github.com/xwp/wordpress-develop/commit/d7c730170770bcea99221d5679fbd12015569214#diff-a325cc1f63ecc04fdb078079c6ec585bR138). This is determined by the [presence of the description](https://github.com/xwp/wordpress-develop/commit/d7c730170770bcea99221d5679fbd12015569214#diff-a325cc1f63ecc04fdb078079c6ec585bR139), so it should be in the same file that produces the description.

westonruter commented:
Great. Let's re-connect on this once 4.6 has kicked off.

kienstra commented:
Thanks, @westonruter! That sounds good.
